codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/UnionGenerator.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.go.codegen;
+
+import software.amazon.smithy.codegen.core.Symbol;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.MemberShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+
+/**
+ * Renders unions and type aliases for all their members.
+ */
+public class UnionGenerator implements Runnable {
+
+    private final Model model;
+    private final SymbolProvider symbolProvider;
+    private final GoWriter writer;
+    private final UnionShape shape;
+
+    UnionGenerator(Model model, SymbolProvider symbolProvider, GoWriter writer, UnionShape shape) {
+        this.model = model;
+        this.symbolProvider = symbolProvider;
+        this.writer = writer;
+        this.shape = shape;
+    }
+
+    @Override
+    public void run() {
+        Symbol symbol = symbolProvider.toSymbol(shape);
+
+        // Creates the parent interface for the union, which only defines a
+        // non-exported method whose purpose is only to enable satisfying the
+        // interface.
+        writer.openBlock("type $L interface {", "}", symbol.getName(), () -> {
+            writer.write("as$L()", symbol.getName());
+        }).write("");
+
+        // Create type aliases for each member that satisfy the interface.
+        for (MemberShape member : shape.getAllMembers().values()) {
+            Symbol memberSymbol = symbolProvider.toSymbol(member);
+            String aliasName = symbol.getName() + symbolProvider.toMemberName(member);
+
+            writer.write("type $L $T", aliasName, memberSymbol)
+                    .write("")
+                    .write("func (v $L) as$L() {}", aliasName, symbol.getName())
```

Comment:
We didn't get a chance to discuss it Friday afternoon, but I believe we might want to rethink this to be a pointer recviever method, rather then a value reciever. This would simplify type switches as it would force the values stored in the interface to be a pointer value. This would inversly complicate  assignment for most types. Similar to how we need methods like `aws.String(string)` for returning pointer types. We could generate `ToPtr()` methods for returning pointers to these type aliases to solve this.